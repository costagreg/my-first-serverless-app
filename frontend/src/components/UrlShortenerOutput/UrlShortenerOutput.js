import React, { useRef, useState } from 'react'
import classNames from 'classnames'

import './UrlShortenerOutput.scss'

const UrlShortenerOutputLink = ({ urlShortened }) => {
  const urlShortenedRef = useRef()
  const [hasCopySuccessed, setHasCopySuccessed] = useState(false)

  const copyToClipboard = (e) => {
    if (urlShortened) {
      urlShortenedRef.current.select()
      document.execCommand('copy')
      urlShortenedRef.current.focus()
      setHasCopySuccessed(true)
    }
  }

  return (
    <>
      <textarea
        className={classNames('urlShortenerOutput__link')}
        ref={urlShortenedRef}
        onClick={copyToClipboard}
        value={urlShortened}
        readOnly
      />
      {hasCopySuccessed && (
        <span className="urlShortenerOutput__copiedMessage">
          Copied to clipboard!
        </span>
      )}
    </>
  )
}

export default ({ urlShortened, error }) => (
  <div
    className={classNames('urlShortenerOutput', {
      'urlShortenerOutput--error': error,
    })}
  >
    {error ? error : <UrlShortenerOutputLink urlShortened={urlShortened} />}
  </div>
)
