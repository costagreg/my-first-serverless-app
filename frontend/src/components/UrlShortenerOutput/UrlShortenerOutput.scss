.urlShortenerOutput {
  text-align: center;

  &__link {
    border: none;
    overflow: auto;
    outline: none;
    -webkit-box-shadow: none;
    -moz-box-shadow: none;
    box-shadow: none;
    resize: none;
    text-decoration: underline;
    text-decoration-style: dotted;
    color: #005073;
    width: 100%;
    text-align: center;
  }

  &__copiedMessage {
    display: block;
  }

  &--error {
    color: #B22222;
  }
}