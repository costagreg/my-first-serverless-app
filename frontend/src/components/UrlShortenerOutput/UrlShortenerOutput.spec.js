import React from 'react'
import UrlShortenerOutput from './UrlShortenerOutput'
import { render, fireEvent, waitFor } from '@testing-library/react'

describe('UrlShortenerOutput', () => {
  describe('@renders', () => {
    it('renders UrlShortenerOutput without any error', () => {
      const { asFragment } = render(<UrlShortenerOutput />)

      expect(asFragment()).toMatchSnapshot()
    })

    describe('error prop is passed', async () => {
      const { getByText } = render(
        <UrlShortenerOutput error="ops" urlShortened="http://www.example.com" />
      )

      const errorMessage = getByText('ops')

      expect(errorMessage).toHaveClass('urlShortenerOutput--error')
    })
  })

  describe('User click in the link', () => {
    beforeEach(() => {
      global.document.execCommand = jest.fn(() => {})
    })

    it('should copy to clipboard', async () => {
      const { getByText } = render(
        <UrlShortenerOutput urlShortened="http://www.example.com" />
      )

      const textLink = getByText('http://www.example.com')
      
      fireEvent.click(textLink)

      await waitFor(() => getByText('Copied to clipboard!'))

      expect(global.document.execCommand).toHaveBeenCalledTimes(1)
      expect(global.document.execCommand).toHaveBeenCalledWith('copy')
    })
  })
})
