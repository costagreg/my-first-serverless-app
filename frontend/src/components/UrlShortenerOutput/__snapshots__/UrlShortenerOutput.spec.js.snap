// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UrlShortenerOutput @renders renders UrlShortenerOutput without any error 1`] = `
<DocumentFragment>
  <div
    class="urlShortenerOutput"
  >
    <textarea
      class="urlShortenerOutput__link"
    />
  </div>
</DocumentFragment>
`;
